### ISSUE TITLE
Support singularity --fusemount with unprivileged user namespace fuse mounts

### ISSUE THREAD

**Issue by DrDaveD** [apptainer/singularity#5206]
Currently singularity --fusemount requires setuid.  However it should not be necessary on newer kernels >= 4.18 that support unprivileged user namespace fuse mounts, and on RHEL >= 7.8.

**Comment by DrDaveD** [comment]
Reproduce on a CentOS 7.8 system with sysctl max_user_namespace > 0 and singularity in the path but `allow setuid = no` in singularity.conf:
```
$ git clone https://github.com/cvmfs-contrib/cvmfsexec.git
$ cd cvmfsexec
$ ./makedist -s osg
$ SINGCVMFS_REPOSITORIES=cms.cern.ch SINGCVMFS_IMAGE=docker://centos:7 ./singcvmfs
INFO:    Convert SIF file to sandbox...
WARNING: underlay of /etc/auto.cvmfs required more than 50 (103) bind mounts
WARNING: underlay of /usr/sbin/mount.cvmfs required more than 50 (113) bind mounts
WARNING: underlay of /usr/bin/cvmfs2 required more than 50 (419) bind mounts
WARNING: underlay of /usr/lib64/libcvmfs_fuse.so required more than 50 (187) bind mounts
WARNING: underlay of /usr/lib/systemd/system/autofs.service.d/50-cvmfs.conf required more than 50 (164) bind mounts
INFO:    Cleaning up image...
FATAL:   container creation failed: mount fuse->/var/singularity/mnt/session/fuse/0 error: while mounting fuse: can't mount fuse filesystem to /var/singularity/mnt/session/fuse/0: invalid argument
```

**Comment by DrDaveD** [comment]
The problem exists in 3.5.3, but I'm not sure it exists in the master branch.  Still testing.

**Comment by cclerget** [comment]
It was fixed in #4953 and is currently in master branch. The `invalid argument` is returned because in 3.5.3 and before the fuse file descriptor is open outside of container user namespace and can't be used for mount if the process doing the call is not inside the user namespace where the file was open, #4953 fix that by reopening fuse file descriptors inside the container user namespace

**Comment by DrDaveD** [comment]
Ok, very good.

I'm experiencing different issues on the master branch, however, and am having trouble figuring out what's happening.  When running non-setuid on CentOS 7.8, I'm getting a warning printed from libfuse3:
```
fuse: failed to clone device fd: Inappropriate ioctl for device
fuse: trying to continue without -o clone_fd.
```
This option is being passed to libfuse3 by cvmfs as a `1` in the second parameter to `fuse_session_loop_mt()` and it is supposed to enhance performance.  It still works, but presumably performance is reduced.  I would suspect that this is just something that isn't supported by the kernel in unprivileged user namespaces, except that I don't get that error when I use cvmfs and libfuse3 from my [cvmfsexec](https://github.com/cvmfs-contrib/cvmfsexec) command.  That command doesn't use premounting, it just lets libfuse3 do its own opening of /dev/fuse and mounting.

On CentOS 8.1 I'm having even more trouble with the --fusemount option, in that with both setuid and non-setuid it is attempting to run fusermount3, which it isn't supposed to do with a pre-mounted file descriptor.
```
fuse: failed to exec fusermount3: No such file or directory
```

If you have any ideas for me let me know, otherwise I'll have to try to debug further and narrow it down to a set of instructions where it is easier to reproduce.

Attn: @jblomer

**Comment by cclerget** [comment]
@DrDaveD If you get `fuse: failed to clone device fd: Inappropriate ioctl for device` it's because `FUSE_DEV_IOC_CLONE` is not supported Centos 7.8 kernel. This feature was added in 4.2 upstream kernel.

Looks like Centos 8.1 is using libfuse 3.2.1 while `/dev/fd/x` handling was added in 3.3.0 explaining why it keep using `fusermount3`

**Comment by DrDaveD** [comment]
> Looks like Centos 8.1 is using libfuse 3.2.1 while `/dev/fd/x` handling was added in 3.3.0 explaining why it keep using `fusermount3`

Thanks for noticing that.  What a pain!  Because 8.1 added support for fuse3, it had to be removed from EPEL.  For now the best I can think of is to use EPEL7 fuse3-libs on EL8.

> If you get `fuse: failed to clone device fd: Inappropriate ioctl for device` it's because `FUSE_DEV_IOC_CLONE` is not supported Centos 7.8 kernel. This feature was added in 4.2 upstream kernel.

Thanks for that pointer too, it does explain the failure to clone.  However, there's a little more to the story.  I don't see the warning on cvmfsexec because then cvmfs is run with default options so it daemonizes itself and in the process replaces stderr with /dev/null.  I ran an strace on it and found that the "failed to clone device fd" was being written but not seen on the terminal.  On the other hand, singularity passes the "-f" option which tells cvmfs to run in the foreground where it does not replace stderr.  This can be avoided by using a prefix of "container-daemon:" instead of "container".  That actually seems to work quite well, but the problem is that it doesn't work in versions of singularity older than the master branch (of course when run with setuid).

So the master branch changes the behavior of "container"; how about if you instead make "container" be backward compatible to not set the "-f", and have a "container-foreground" to add "-f" instead of a "container-daemon" to take it away?

Another difference between singularity-3.5.3's fusemount and the master branch's fusemount is that the former shows no output at all from cvmfs.  I don't mind the difference, however, as long as cvmfs isn't run with -f.

**Comment by DrDaveD** [comment]
@cclerget What do you think about changing the "container-daemon" behavior to be the default?

**Comment by cclerget** [comment]
@DrDaveD My opinion on this is to left `container` as the default behavior, it doesn't matter for CVMFS use case as it's a RO filesystem but doing that with RW filesystem may cause filesystem corruption, in master branch only FUSE processes running in foreground mode are receiving SIGTERM signal at container exit because sinit process can track/wait them easily, FUSE processes running as daemon are another story and for simplicity of code those processes are considered as orphaned and therefore are killed with SIGKILL signal by the kernel when the pid namespace is destroyed, as the PID namespace is destroyed before the mount namespace there is no chance that umount occurs before being killed which could lead to potential corruption. That's why foreground is the default mode because it's easier to track/manage FUSE processes and it made explicit to request daemon mode with the consequences involved by doing that.

**Comment by dtrudg** [comment]
This discussion probably means it's also time to document fusemount with these caveats. I'll open a docs issue.

@DrDaveD - I'm afraid that @cclerget 's reasoning makes sense to me here - and because the `--fusemount` option has been an explicitly experimental hidden flag I'm not too concerned about a behaviour change on the CLI if it makes the most sense for `--fusemount` now being surfaced.

**Comment by DrDaveD** [comment]
Well it gives me a compatibility problem because there's not one way to do it that will work with old and new versions.  How about another environment variable that reverts it to the old behavior?  It will be ignored by older versions.

**Comment by dtrudg** [comment]
> Well it gives me a compatibility problem because there's not one way to do it that will work with old and new versions. How about another environment variable that reverts it to the old behavior? It will be ignored by older versions.

I'm afraid that I'm against adding additional complexity with an environment variable here. I think in general if we set a precedent of maintaining backwards compatibility for functionality that was marked experimental, and was explicitly an undocumented hidden flag then it will complicate future work.

**Comment by DrDaveD** [comment]
How do you suggest I then make a script work with old & new versions of singularity?

**Comment by DrDaveD** [comment]
> This discussion probably means it's also time to document fusemount with these caveats. I'll open a docs issue.

Make sure it covers the help message, which currently only mentions "container:".

**Comment by dtrudg** [comment]
Hi @DrDaveD -  it's not an absolute no here, but the new behaviour was merged back in Jan 23rd and we are due to issue an RC for 3.6 this week so I don't want to introduce more changes in this area right now. I'll ask others for their opinion on this too.

> How do you suggest I then make a script work with old & new versions of singularity?

The output of `singularity version` can be checked in a script, e.g.:

```
echo "$(singularity version | grep -o '^[[:digit:]].[[:digit:]]')" | awk '$1 > 3.5 {print "new"}'
```

>> This discussion probably means it's also time to document fusemount with these caveats. I'll open a docs issue.

> Make sure it covers the help message, which currently only mentions "container:".

We'll do what we can given the RC and release timeline.

**Comment by DrDaveD** [comment]
Well, RC's are "candidates" and are allowed to change if problems are found. 

I'd really like to avoid a version hack.  I think I'd rather introduce a wrapper script that either removes the -f or sends stdout/stderr to /dev/null by default.  In fact, the extra cvmfs startup messages are annoying, and maybe the solution is for singularity --fusemount to hide them by default, unless a verbose or debug option is given.

And in fact even though cvmfs is read-only it does put a message in /var/log/messages at mount time if it was previously unmounted uncleanly, so it would be nice to get clean unmounts.  I am still seeing those messages, even with the -f, which says to me that the unmounts Cedric talked about are not really happening.  I also don't see any mention of unmounts at exit time with -vddd options.

**Comment by dtrudg** [comment]
> Well, RC's are "candidates" and are allowed to change if problems are found. 

Yes - that was my point with it not being absolute. If there's some consensus we can certainly make changes. Sorry, I didn't want to come across as no changes for 3.6 - just no immediate change before an RC goes out - and then let's look into this a bit more and arrive at a consensus that involves others as well.

Hopefully @cclerget will have some time in the near future to look at the other concern raised too.

Cheers.

**Comment by DrDaveD** [comment]
If I have a very fast command to execute and have debugging enabled, I will often but not always see stopFuserDrivers() calls:
```
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=9402]    PostStartProcess()            Post start process
DEBUG   [U=3382,P=1]       stopFuseDrivers()             FUSE process for mount point /cvmfs/config-osg.opensciencegrid.org terminated
DEBUG   [U=3382,P=1]       stopFuseDrivers()             FUSE process for mount point /cvmfs/atlas.cern.ch terminated
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=12564]   Master()                      Child exited with exit status 0
```
If the command runs longer I don't see that, I just see a lot of "Received signal child exited" at the beginning:
```
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
DEBUG   [U=3382,P=1]       StartProcess()                Received signal child exited
```
and only this when it exits:
```
DEBUG   [U=3382,P=11956]   Master()                      Child exited with exit status 0
```

**Comment by carterpeel** [comment]
Hello,

This is a templated response that is being sent out to all open issues.  We are working hard on 'rebuilding' the Singularity community, and a major task on the agenda is finding out what issues are still outstanding.


**Please consider the following:**

1. Is this issue a duplicate, or has it been fixed/implemented since being added?
2. Is the issue still relevant to the current state of Singularity's functionality?
3. Would you like to continue discussing this issue or feature request?


Thanks,
Carter

**Comment by DrDaveD** [comment]
Fixed in #4953.
